configtype/time_test.go

```diff
@@ -0,0 +1,180 @@
+		{"2021-09-01T00:00:00.123456+02:00", time.Date(2021, 9, 1, 0, 0, 0, 123456000, time.FixedZone("CET", 2*60*60))},
+		{"2021-09-01T00:00:00.123456789+02:00", time.Date(2021, 9, 1, 0, 0, 0, 123456789, time.FixedZone("CET", 2*60*60))},
+		{"2021-09-01", time.Date(2021, 9, 1, 0, 0, 0, 0, time.UTC)},
+		{"now", now},
+		{"2 days from now", now.AddDate(0, 0, 2)},
+		{"5 months ago", now.AddDate(0, -5, 0)},
+	}
+	for _, tc := range cases {
+		var d configtype.Time
+		err := json.Unmarshal([]byte(`"`+tc.give+`"`), &d)
+		if err != nil {
+			t.Fatalf("error calling Unmarshal(%q): %v", tc.give, err)
+		}
+		computedTime := d.Time(now)
+		if !computedTime.Equal(tc.want) {
+			t.Errorf("Unmarshal(%q) = %v, want %v", tc.give, computedTime, tc.want)
+		}
+	}
+}
+
+func TestTime_Comparability(t *testing.T) {
+	tim1 := time.Now()
+	tim2 := tim1.Add(1 * time.Second)
+
+	var zeroTime configtype.Time
+
+	cases := []struct {
+		give    configtype.Time
+		compare configtype.Time
+		equal   bool
+	}{
+		{must(configtype.ParseTime("now")), must(configtype.ParseTime("now")), true},
+		{must(configtype.ParseTime("now")), must(configtype.ParseTime("1 second from now")), false},
+		{configtype.NewTime(tim1), configtype.NewTime(tim1), true},
+		{configtype.NewTime(tim1), configtype.NewTime(tim2), false},
+		// relative and fixed times are never equal
+		{configtype.NewTime(tim1), must(configtype.ParseTime("now")), false},
+		{zeroTime, must(configtype.ParseTime("now")), false},
+		{zeroTime, zeroTime, true},
+	}
+	for _, tc := range cases {
+		if (tc.give == tc.compare) != tc.equal {
+			t.Errorf("comparing %v and %v should be %v", tc.give, tc.compare, tc.equal)
+		}
+
+		diff := cmp.Diff(tc.give, tc.compare)
+		if tc.equal && diff != "" {
+			t.Errorf("comparing %v and %v should be equal, but diff is %s", tc.give, tc.compare, diff)
+		} else if !tc.equal && diff == "" {
+			t.Errorf("comparing %v and %v should not be equal, but diff is empty", tc.give, tc.compare)
+		}
+	}
+}
+
+func TestTime_JSONSchema(t *testing.T) {
+	sc := (&jsonschema.Reflector{RequiredFromJSONSchemaTags: true}).Reflect(configtype.Time{})
+	schema, err := json.MarshalIndent(sc, "", "  ")
+	require.NoError(t, err)
+
+	validator, err := plugin.JSONSchemaValidator(string(schema))
+	require.NoError(t, err)
+
+	type testCase struct {
+		Name string
+		Spec string
+		Err  bool
+	}
+
+	for _, tc := range append([]testCase{
+		{
+			Name: "empty",
+			Err:  true,
+			Spec: `""`,
+		},
+		{
+			Name: "null",
+			Err:  true,
+			Spec: `null`,
+		},
+		{
+			Name: "bad type",
+			Err:  true,
+			Spec: `false`,
+		},
+		{
+			Name: "bad format",
+			Err:  true,
+			Spec: `false`,
+		},
+		{
+			Name: "not relative duration",
+			Err:  true,
+			Spec: `"10 days"`,
+		},
+		{
+			Name: "relative duration",
+			Err:  false,
+			Spec: `"10 months from now"`,
+		},
+		{
+			Name: "complex relative duration",
+			Err:  false,
+			Spec: `"10 months 3 days 4h20m from now"`,
```

Comment:
I think if we're going to expose `MarshalJSON` then we should probably make this format `10 months 3 days 4 hours 20 minutes from now` for consistency (and so we don't need to make a breaking change later to fix it)

Reply:
Alternatively we could remove the ability to specify an arbitrary duration in the constructor (just remove the constructor) and only allow this to be constructed using a simple string like `10 days ago`

Reply:
The reason for the weird hybrid format is for backwards compat for the Duration type. It ensures that anything constructed from a plain duration will emit the same json as before this PR. It may be reasonable to separate the json serialization format and the string parsed format and make the serialization format (but still unmarshalling the nicer human format) just `[-+](\d+[a-zA-Z])+` (basically time.ParseDuration format but with additional units). That frees us to have the string format be nicer and human readable

Reply:
Would https://github.com/cloudquery/plugin-sdk/pull/1905/files#r1775175473 solve this?